DEP-06: Immutable ETCD Backups (#1)
seshachalam-yv/etcd-druid: docs/proposals/06-immutable-etcd-backups.md

---
title: Immutable ETCD Backups
dep-number: 06
creation-date: 2024-09-25
status: implementable
authors:
- "@seshachalam-yv"
- "@renormalize"
- "@ishan16696"
reviewers:
- "@unmarshall"
---

# DEP-06: Immutable ETCD Backups

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
- [Proposal](#proposal)
    - [Overview](#overview)
    - [Detailed Design](#detailed-design)

>         - [Bucket Lock Mechanism](#bucket-lock-mechanism)

Reviewer: The term `Bucket Lock` is identical to the terminology used by one vendor - GCS. Perhaps we could come up with terminology which is vendor agnostic, since we plan on supporting multiple and each has their own terminology?

    - [Garbage Collection during Compaction for Hibernated ETCD Clusters](#garbage-collection-during-compaction-for-hibernated-etcd-clusters)
- [Compatibility](#compatibility)
- [Implementation Steps](#implementation-steps)
- [Risks and Mitigations](#risks-and-mitigations)
- [Operational Considerations](#operational-considerations)
- [Alternatives](#alternatives)
    - [Object-Level Retention Policies vs. Bucket-Level Retention Policies](#object-level-retention-policies-vs-bucket-level-retention-policies)
        - [Feasibility Study: Immutable Backups on Cloud Providers](#feasibility-study-immutable-backups-on-cloud-providers)
        - [Considerations for Object-Level Retention](#considerations-for-object-level-retention)
        - [Conclusion](#conclusion)
- [References](#references)

## Summary

This proposal aims to enhance the reliability and integrity of ETCD backups by introducing immutable backups. By leveraging cloud provider features that support a write-once-read-many (WORM) model, this approach prevents unauthorized modifications to backup data, ensuring that backups remain available and intact for restoration.

## Motivation

> Ensuring the integrity and availability of ETCD backups is crucial for the stability and reliability of ETCD clusters. Making backups immutable protects against unintended or malicious modifications post-creation, thereby enhancing the overall security posture.

Reviewer: The backups don't have anything to do with the stability of *currently running* etcd clusters. Only in cases of restoration do we care about the integrity of the backups. So perhaps we could emphasize that integrity matters whenever we need to restore?

Also do we need to specify "security posture"?
```suggestion
Ensuring the integrity and availability of ETCD backups is crucial for the ability to restore an ETCD cluster after it has irrecoverably gone down. Making backups immutable protects against unintended or malicious modifications post-creation, thereby enhancing the overall security posture.
```

### Goals

- Implement immutable backup support for ETCD clusters.
- Secure backup data against unintended or unauthorized modifications after creation.
- Ensure backups are consistently available and intact for restoration purposes.

### Non-Goals

- Altering existing backup processes beyond what's necessary for immutability.
- Implementing object-level retention policies at this stage.

## Proposal

### Overview

Introduce immutability in backup storage by leveraging cloud provider features that support a write-once-read-many (WORM) model. This will prevent data alterations after backup creation, enhancing data integrity and security.

### Detailed Design

#### Bucket Lock Mechanism

> The Bucket Lock feature configures a retention policy for a cloud storage bucket, governing how long objects in the bucket must be retained. It also allows for locking the bucket's retention policy, permanently preventing the policy from being reduced or removed.

Reviewer: The phrase `retention policy` might be confusing for someone who is familiar with the etcd repos in gardener. Case in point: https://github.com/gardener/etcd-backup-restore/pull/776#discussion_r1766294499.

In etcd-backup-restore, we already have a feature called `delta snapshot retention period` which might be confused with what is being expressed here - the duration for which a snapshot is immutable after upload.

Can we use something like `immutability period` as expressed here https://github.com/gardener/etcd-backup-restore/pull/776#discussion_r1766764425?

- **Supported by Major Providers:**
    - **Google Cloud Storage (GCS):** [Bucket Lock](https://cloud.google.com/storage/docs/bucket-lock)
    - **Amazon S3 (S3):** [Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html)
    - **Azure Blob Storage (ABS):** [Immutable Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-policy-configure-container-scope?tabs=azure-portal)

- **Not Supported:**
    - **OpenStack Swift**

> Operators are responsible for creating or updating buckets with these immutable settings. Etcd-druid will use the bucket details from `Etcd.Spec.Backup.Store.Container` to upload the ETCD snapshots.

Reviewer: Makes it more explicit? I don't mind if you don't want this extra sentence tbf.
```suggestion
Operators are responsible for creating or updating buckets with these immutable settings. Etcd-druid will continue to use the bucket details from `Etcd.Spec.Backup.Store.Container` to upload the ETCD snapshots. No additional configuration needs to be passed to etcd-druid.
```

Once the bucket is configured to be immutable, snapshots uploaded by `etcd-backup-restore` are also immutable.

#### ETCD Backup Configuration

> Operators need to ensure that the ETCD backup configuration aligns with the immutability requirements. This includes setting appropriate retention periods.

Reviewer: Retention period here is slightly confusing again.

> #### Handling of Hibernated Clusters

Reviewer: Since hibernation is a gardener/gardener concept, could we specify that in contexts where consumers/operators like gardener/gardener bring down their etcd clusters, the immutability guarantees might be at risk?

In scenarios where an ETCD cluster has been hibernated for a duration that exceeds the immutable retention period, backups may become mutable again, compromising the intended immutability guarantees.

There are two ways to maintain the immutability of snapshots:

1. **Extend the Immutability of Snapshots:** With bucket-level immutable settings, we cannot extend the immutable retention period for a specific snapshot since it is a bucket-level configuration.

2. **Take a New Snapshot:** By creating a new snapshot, immutability is retained for the new snapshot.

> Approach 2 is feasible and can be achieved in two ways:

Reviewer: Can we also mention how approach 1 is not feasible at all since it would increase the immutability duration for **all** objects that are uploaded, including the ones that will be uploaded in the future to a longer duration; would significantly drive storage costs, so on?

> - **Option A:** Wake up the ETCD cluster by increasing the replicas, take a snapshot, and then hibernate the ETCD by setting the replicas to zero.

Reviewer: This assumes the reader knows what happens in gardener/gardener's hibernation. Maybe we could instead say that the etcd cluster is brought up?

- **Option B:** Start an embedded ETCD instance and take a snapshot.

Option A is complicated because we cannot increase the ETCD replica size without operator intervention.

> Option B is recommended since we are not disturbing the existing setup, and we can leverage the compaction job to take responsibility for this.

Reviewer: Just a suggestion, feel free to discard.
```suggestion
Option B is recommended since this is orthogonal to the existing setup, and we can leverage the compaction job to take responsibility for this.
```

**Proposed Solution:**

> Leverage the compaction job to take fresh snapshots periodically during hibernation, thus ensuring that new, immutable snapshots are regularly created and the integrity of the backup immutability is maintained.

Reviewer: ```suggestion
Leverage the compaction job to take fresh snapshots periodically during hibernation, thus ensuring that new, immutable snapshots are regularly created and the integrity of the backups is maintained.
```

    - **Mitigation:** Carefully configure the retention period to balance between data protection needs and storage cost considerations.

- **Risk:** Hibernated clusters might not receive necessary backups, potentially leading to compliance issues with the retention policy.
    - **Mitigation:** Implement the recommended proactive snapshotting approach to ensure backups are taken even during hibernation.

- **Risk:** Excluding snapshots during restoration might be misused or lead to incomplete data restoration.
    - **Mitigation:** Restrict the ability to tag snapshots for exclusion to authorized personnel and ensure proper audit logging is in place.

## Operational Considerations

Operators need to:

- Configure buckets with appropriate immutability settings.
- Ensure that the compaction job is properly configured to handle snapshotting during hibernation.
- Implement policies and procedures for excluding snapshots when necessary.
- Monitor storage costs due to the potential increase from immutable backups.
- Keep track of the retention periods to ensure compliance.

## Alternatives

### Object-Level Retention Policies vs. Bucket-Level Retention Policies

> An alternative to implementing immutability via bucket-level retention is to use object-level retention policies. Object-level retention allows for more granular control over the retention periods of individual objects within a bucket, whereas bucket-level retention applies a uniform retention period to all objects in the bucket.

Reviewer: Same suggestions about retention period v/s immutability period apply here specifically.

<details><summary>Comparison of Storage Provider Properties for Bucket-Level and Object-Level Immutability</summary>

| Feature                                                           | GCS | AWS | Azure |
|-------------------------------------------------------------------|-----|-----|-------|
| Can bucket-level retention period be increased?                   | Yes | Yes* | Yes (only 5 times) |
| Can bucket-level retention period be decreased?                   | No  | Yes* | No    |
| Is bucket-level retention a prerequisite for object-level retention? | No  | Yes | Yes (for existing buckets), No (for new buckets) |
| Can object-level retention period be increased?                   | Yes | Yes | Yes   |
| Can object-level retention period be decreased?                   | No  | No  | No    |
| Support for enabling object-level immutability in existing buckets | No (planned support soon) | Yes (only new objects will have immutability) | Yes (Azure handles the migration) |
| Support for enabling object-level immutability in new buckets     | Yes | Yes | Yes   |

> | Precedence between bucket-level and object-level retention periods | Maximum of bucket or object-level retention | Object-level retention has precedence | Object-level retention has precedence |

Reviewer: ```suggestion
| Precedence between bucket-level and object-level retention periods | Maximum of bucket or object-level retention | Object-level retention has precedence | Maximum of bucket or object-level retention |
```
